### DISCUSSION TITLE
bash/sh/shell codeblock syntax highlight

### DISCUSSION THREAD

**Discussion by GSB-Deleven** [squidfunk/mkdocs-material#6504] [Q&A]
Hi,

in VS Code my bash/sh/shell codeblocks look like this:
![image](https://github.com/squidfunk/mkdocs-material/assets/35196079/cf08e359-2973-4827-9610-1dd4fc8d1c8b)

is there a way to do this in mkdocs-material?

Currently it renders like this:
![image](https://github.com/squidfunk/mkdocs-material/assets/35196079/072e3cf9-dde9-41aa-b267-dfeabeea6b0a)

Thanks a lot for your help

**Comment by alexvoss** [comment]
This is really a question about the syntax highlighter used. Note that you have [two options there, Pygments or a JavaScript highlighter](https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#highlight), with Pygments being the default. The Pygments support for bash seems somewhat limited. 

You may have more luck using a JavaScript highlighter. Having said that, [highlight.js](https://highlightjs.org) did not fare better with a mount command. I guess they would all work better with actual bash syntax such as for loops, conditionals, etc.?

To be honest, I personally do not find either version particularly helpful. The more colorful rendering that VSCode produces does not exactly clarify things, I think (why are `-t` and `192.168` both orange?).

**Reply by GSB-Deleven** [reply]
Thanks for the super quick response
happy to see that it isn't something I missed.

You are right, the 192.168 in orange doesn't make sense for me either. 
(But I actually like the -t to separate flags)

**Reply by kamilkrzyskow** [reply]
Searching up what does VS Code use I came across TextMate.
There is a node tool which allows to use the tmlanguage files to highlight text https://github.com/ibilon/Highlighter

But I'm not sure how to use it in a browser setting instead of the highlight.js module.

There is also the option of writing your own lexer for Pygments with some custom regex.
Tough keep in mind that currently custom lexers aren't cached in the build process and might significantly decrease docs build performance given a lot of code examples.

**Reply by squidfunk** [reply]
IIRC, @hellt wrote a custom Pygments Lexer, and it was quite straight forward, wasn't it?

**Reply by hellt** [reply]
it was moderately straightforward, but still some scripting skills required to put it all together.
in case folks want to get some headstart I documented my journey [here](https://learn.srlinux.dev/blog/2023/sr-linux-syntax-highlighting-with-pygments).

**Reply by squidfunk** [reply]
Awesome, thanks for sharing!

**Reply by kamilkrzyskow** [reply]
I wrote a Pygments lexer myself https://github.com/kamilkrzyskow/gothic-lexer for the Daedalus game scripting language similar to C/C++
Once I finally understood the "sequential tokenization" part, where the root level context switches to function level context and then goes out of it (understanding that part took me a good month or so), then the lexer was done in a week or 2.

@GSB-Deleven can also modify the `BashLexer` https://github.com/pygments/pygments/blob/master/pygments/lexers/shell.py#L25 as it's probably a good base to start. Or perhaps try the different `BashSessionLexer` with the `console` instead of `sh`, as it seems to be designed with commands in mind. https://pygments.org/docs/lexers/#module-pygments.lexers.shell

**Reply by GSB-Deleven** [reply]
Thank you very much to all of you.
I'll read in to it and try to figure it out.

Great to have such an active and helpful community/developer

**Comment by seub** [comment]
Hi all, just adding that I also find the shell syntax highlighting unsatisfying, for example:

<img width="796" alt="Screenshot 2024-09-03 at 6 21 25 PM" src="https://github.com/user-attachments/assets/449e0d45-f324-4a89-8f35-707765c12e81">

Those colors do more harm than good.

**Reply by squidfunk** [reply]
This should be reported to [Pygments](https://github.com/pygments/pygments), which does the syntax highlighting. We only provide the coloring 😅

**Reply by seub** [reply]
Got it. Looks like this issue was already reported [here](https://github.com/pygments/pygments/issues/2668)

Are there alternative highlighters that we could use?

**Reply by squidfunk** [reply]
Pygments is the default and it's likely currently not easy to switch to something different in the Python ecosystem, but you might be able to use a JavaScript-based highlighter like Highlight.js. Note that we do not officially support those, so some features like code annotations will currently not work when deviating from the defaults.

**Reply by squidfunk** [reply]
Also note that you can just use `txt` as a language to remove the highlighting, which might be easier.

**Reply by seub** [reply]
Thanks!

**Reply by facelessuser** [reply]
90% of the time something like `console` (`shell-session`, etc.) language works fine for me because I'm not triggering special Bash specific tokens.

<img width="789" alt="Screenshot 2024-09-04 at 8 46 48 AM" src="https://github.com/user-attachments/assets/ebf744bf-f1d8-42d1-8576-9f417e9d7356">

But when you trigger certain Bash tokens, it can try to colorize them and be a bit surprising. In the example `pyenv local 3.11.15` if you quote the number, it looks okay, but it adds unnecessary steps:

<img width="425" alt="Screenshot 2024-09-04 at 8 48 09 AM" src="https://github.com/user-attachments/assets/b664b09d-578d-421f-b271-6110fc6509c2">

An easy way to overcome this is maybe to write your own simple terminal lexer that highlights the user prompt, entire command, and result as different things, similar to what I showed above but without the Bash nuance.  It's not too hard, but I imagine most people don't want to go down that route. I've done some in the past: https://github.com/facelessuser/pymdown-lexers.

Plain texts for a simple command is often sufficient for most people. I can see the lexer approach being more useful if you are often showing actual command prompts with responses and you'd like a visual way to separate the response from the command and avoid the Bash specific nuance.

**Reply by seub** [reply]
@facelessuser Thank you for the comment and sharing your project! It seems to me solutions such as "quote the numbers" or "use txt" are not ideal, because they don't reflect the intent. I'd hope one day Pygments fixes their highlighting, or I'd be able to choose another highlighter/lexer, and I don't want to remember all the places where I'll need to switch back.

**Reply by facelessuser** [reply]
To be honest, Bash is a difficult language to highlight correctly. I don't find the VS code example in the opening post much better than the one in MkDocs Material. I personally think a lexer that ignores the nuance of the shell and can target the prompt, user input, and result would be much better, but even in that case, you could still have false positives if the result shows things that mimic the input, but I think it would be better. I don't think you are likely to see a perfect highlighter of such outputs, but I think a more generic one would be better. I don't know if you'll ever see that in Pygments, but who knows.

**Reply by hellt** [reply]
@facelessuser thanks for the pymdown-lexers, it helped me to understand how to install the lexers in the mkdocs project when I built a custom lexer.
I documented the whole process here https://learn.srlinux.dev/blog/2023/sr-linux-syntax-highlighting-with-pygments


with the flexibility of customer lexers the biggest features I find unsatisfied with mkdocs-material and the surrounding ecosystem are

1. expandable code blocks (full screen view)
2. configurable copy ranges (this is to copy the payload, and not the prompt with it)

**Reply by facelessuser** [reply]
@hellt Glad you found the lexer project useful. It can be a little confusing when first attempting such things, so it's always nice to have a simple reference. It's awesome that you provided a write up.
